File: 4_Roman2Int/solution.py

def romanToInt(s):
    """
    :type s: str
    :rtype: int
    """
    dict_values = {"I": 1, "V": 5, "X": 10, "L":50,"C":100,"D":500,"M":1000}
    res = 0
    i=0
    while i < len(s):
        

        if( i < len(s)-1 and ((s[i] == 'I' and s[i+1] == 'V') or( s[i] == 'I' and s[i+1] == 'X'))):
            res += dict_values[s[i+1]] - dict_values[s[i]] 
            i = i + 2
            continue
        
        elif(i < len(s)-1 and  ((s[i] == 'X' and s[i+1] == 'L' )or (s[i] == 'X' and s[i+1] == 'C'))):
            res += dict_values[s[i+1]] - dict_values[s[i]] 
            i = i + 2
            continue
        
        elif(i < len(s)-1 and  ((s[i] == 'C' and s[i+1] == 'D') or( s[i] == 'C' and s[i+1] == 'M' ))):
            res += dict_values[s[i+1]] - dict_values[s[i]] 
            i = i + 2
            continue
        

        res = res + dict_values[s[i]]
        i= i + 1

    return res


res = romanToInt(s)
print("resutl: " , res)

File: 4_Roman2Int/solution.cpp
#include <iostream>
#include <map>
#include <string>

int romanToInt(std::string s ){
    std::map<std::string, int> myMap;

    myMap["I"] = 1;
    myMap["V"] = 5;
    myMap["X"] = 10;
    myMap["L"] = 50;
    myMap["C"] = 100;
    myMap["D"] = 500;
    myMap["M"] = 1000;

    int res = 0;
    int i = 0;

    while( i < s.length()){

        if (i < s.length()-1 && (s[i] == 'I' &&( s[i+1] == 'V' || s[i+1] == 'X'))){
            res +=  myMap[std::string(1,s[i+1])] -  myMap[std::string(1,s[i])];
            i += 2;
            continue;
        }

        if (i < s.length()-1 && (s[i] == 'X' && (s[i+1] == 'L' || s[i+1] == 'C'))){
            res +=  myMap[std::string(1,s[i+1])] -  myMap[std::string(1,s[i])];
            i += 2;
            continue;
        }

        if (i < s.length()-1 && (s[i] == 'C' && (s[i+1] == 'D' || s[i+1] == 'M'))){
            res +=  myMap[std::string(1,s[i+1])] -  myMap[std::string(1,s[i])];
            i += 2;
            continue;
        }

        res += myMap[std::string(1,s[i])];
        i += 1;

        
    }

    return res;
}


int main() {
    
    int res = romanToInt("MCMXCIV");
    std::cout << res << std::endl;

    return 0;
}